kerastuner/engine/metrics_tracking.py

```diff
@@ -22,136 +22,220 @@
 from tensorflow import keras
 
 
-MetricObservation = collections.namedtuple(
-    'MetricObservation',
-    'value step')
+class Direction(object):
+    def __init__(self, direction='min'):
+        if direction not in {'min', 'max'}:
+            raise ValueError(
+                '`direction` should be one of '
+                '{"min", "max"}, but got: %s' % (direction,))
+        self.direction = direction
+
+    def get_best(self, values):
+        if not values:
+            return None
+        if self.direction == 'min':
+            return np.nanmin(values)
+        return np.nanmax(values)
+
+    def get_config(self):
+        return {'direction': self.direction}
+
+    @classmethod
+    def from_config(cls, config):
+        return cls(config['direction'])
+
+
+class MetricObservation(object):
+
+    def __init__(self, value, step):
+        self.value = value
+        self.step = step
+
+    def append(self, value):
+        if not isinstance(value, list):
+            value = [value]
+        if not isinstance(self.value, list):
+            self.value = [self.value]
+        self.value += value
+
+    def mean(self):
+        if not isinstance(self.value, list):
+            return self.value
+        return np.mean(self.value)
+
+    def get_config(self):
+        return {'value': self.value,
+                'step': self.step}
+
+    @classmethod
+    def from_config(cls, config):
+        return cls(**config)
+
+    def __eq__(self, other):
+        if not isinstance(other, MetricObservation):
+            return False
+        return (other.value == self.value and
+                other.step == self.step)
+
+
+class MetricTracker(object):
+
+    def __init__(self, direction='min'):
+        self._direction = Direction(direction)
+        self._observations = {}
+
+    @property
+    def direction(self):
+        return self._direction.direction
+
+    @direction.setter
+    def direction(self, direction):
+        self._direction = Direction(direction)
+
+    def update(self, value, step):
+        if step in self._observations:
+            self._observations[step].append(value)
+        else:
+            self._observations[step] = MetricObservation(
+                value, step=step)
+
+    def get_best_value(self):
```

Comment:
The one method of `Direction` is only used here. This is where we should be doing the sorting.

Reply:
Done